runtime/src/accounts_index.rs

```diff
@@ -204,30 +204,94 @@ impl<T: 'static + Clone + IsCached> WriteAccountMapEntry<T> {
         })
     }
 
+    fn addref(item: &AtomicU64) {
+        item.fetch_add(1, Ordering::Relaxed);
+    }
+
+    pub fn upsert_new_key<'a>(
+        mut w_account_maps: AccountMapsWriteLock<'a, T>,
+        pubkey: &Pubkey,
+        new_value: AccountMapEntry<T>,
+        reclaims: &mut SlotList<T>,
+    ) {
+        match w_account_maps.entry(*pubkey) {
+            Entry::Occupied(mut occupied) => {
+                let current = occupied.get_mut();
+                Self::upsert_item(current, new_value, reclaims);
+            }
+            Entry::Vacant(vacant) => {
+                vacant.insert(new_value);
+            }
+        }
+    }
+
+    // returns None if upsert was successful
+    // otherwise, the pubkey has to be added to the maps with a write lock. call upsert_new
+    pub fn upsert_existing_key<'a>(
+        r_account_maps: AccountMapsReadLock<'a, T>,
+        pubkey: &Pubkey,
+        new_value: AccountMapEntry<T>,
+        reclaims: &mut SlotList<T>,
+    ) -> Option<AccountMapEntry<T>> {
```

Comment:
It seems more intuitive if this returned a boolean, true for success, false if a new entry needs to be created. `new_value` will have to be passed by reference though

Reply:
It is clearer. Since this new_value is still a RwLock, I can do this with a reference without incurring an unnecessary copy. The jury is still out on whether we retain a RwLock or not. Master still has it, so it makes sense to make a simpler api here. This may have to be revisited later. Compile errors will result, yay rust!